File: agent_memory_mcp/correlator.py
"""Memory correlator: finds relationships between memories across repos."""

from __future__ import annotations

import json
import re
from typing import TYPE_CHECKING, Any, Dict, List, Optional

if TYPE_CHECKING:
    from .engine.db import MemoryEngine


class MemoryCorrelator:
    """Finds related memories using content overlap, file references, and temporal proximity."""

    def __init__(self, engine: "MemoryEngine"):
        self.engine = engine

    def get_related(self, memory_id: str, limit: int = 10) -> List[Dict[str, Any]]:
        """Find memories related to the given memory."""
        rows = self.engine._conn.execute(
            "SELECT * FROM memories WHERE id = ? AND deleted = 0", (memory_id,)
        ).fetchall()
        if not rows:
            return []
        source = dict(rows[0])

        candidates: Dict[str, Dict[str, Any]] = {}

        # Strategy 1: Same file references
        file_refs = self._extract_file_refs(source["content"])
        if file_refs:
            for ref in file_refs[:3]:
                hits = self.engine.search(ref, limit=limit * 2)
                for hit in hits:
                    if hit["id"] != memory_id:
                        score = candidates.get(hit["id"], {}).get("relevance", 0)
                        candidates[hit["id"]] = {
                            **hit,
                            "relevance": score + 0.4,
                            "reason": "shared_file",
                        }

        # Strategy 2: Content similarity via search
        query_text = source["content"][:300]
        similar = self.engine.search(query_text, limit=limit * 2)
        for hit in similar:
            if hit["id"] != memory_id:
                score = candidates.get(hit["id"], {}).get("relevance", 0)
                sim_score = hit.get("score", 0.3)
                candidates[hit["id"]] = {
                    **hit,
                    "relevance": score + sim_score,
                    "reason": "content_similar",
                }

        # Strategy 3: Temporal proximity (within 5 minutes)
        created = source["created_at"]
        temporal = self.engine._conn.execute(
            """
            SELECT id, repo_id, kind, content, source, metadata_json, session_id,
                   summary, created_at
            FROM memories
            WHERE deleted = 0 AND id != ?
            AND abs(julianday(created_at) - julianday(?)) < (5.0 / 1440.0)
            ORDER BY created_at DESC LIMIT ?
            """,
            (memory_id, created, limit * 2),
        ).fetchall()
        for row in temporal:
            mid = row[0]
            if mid not in candidates:
                candidates[mid] = {
                    "id": mid,
                    "repo_id": row[1],
                    "kind": row[2],
                    "content": row[3],
                    "source": row[4],
                    "metadata": json.loads(row[5]) if row[5] else {},
                    "session_id": row[6],
                    "summary": row[7],
                    "created_at": row[8],
                    "relevance": 0.2,
                    "reason": "temporal_proximity",
                }
            else:
                candidates[mid]["relevance"] += 0.2

        ranked = sorted(
            candidates.values(), key=lambda x: x.get("relevance", 0), reverse=True
        )
        return ranked[:limit]

    def _extract_file_refs(self, text: str) -> List[str]:
        """Extract file path references from text."""
        patterns = [
            r"[\w/.-]+\.\w{1,5}",
            r"(?:src|lib|app|tests?)/[\w/.-]+",
        ]
        refs: set[str] = set()
        for pattern in patterns:
            for match in re.finditer(pattern, text):
                ref = match.group()
                if len(ref) > 5 and not ref.startswith("http"):
                    refs.add(ref)
        return list(refs)[:10]

    def build_dependency_graph(
        self, repo_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """Build a graph of memory relationships for visualization."""
        memories = self.engine.list_memories(repo_id=repo_id, limit=100)

        nodes = []
        edges = []

        for mem in memories:
            nodes.append(
                {
                    "id": mem["id"],
                    "kind": mem["kind"],
                    "repo_id": mem["repo_id"],
                    "label": (mem.get("summary") or mem["content"][:60]),
                    "created_at": mem["created_at"],
                }
            )

            related = self.get_related(mem["id"], limit=3)
            for rel in related:
                if rel.get("relevance", 0) > 0.3:
                    edges.append(
                        {
                            "source": mem["id"],
                            "target": rel["id"],
                            "weight": rel.get("relevance", 0),
                            "reason": rel.get("reason", "unknown"),
                        }
                    )

        return {"nodes": nodes, "edges": edges}

File: internal/patterns/correlator.go
package patterns

import (
	"context"
	"regexp"
	"sort"

	"agent-memory-mcp/internal/core"
	"agent-memory-mcp/internal/store/sqlite"
)

// Correlator finds relationships between memories using content overlap,
// file references, and temporal proximity.
type Correlator struct {
	store *sqlite.Store
}

func NewCorrelator(store *sqlite.Store) *Correlator {
	return &Correlator{store: store}
}

// GetRelated finds memories related to the given memory.
func (c *Correlator) GetRelated(ctx context.Context, memoryID string, limit int) ([]core.Memory, error) {
	// Get the source memory
	mems, err := c.store.ListMemories(ctx, nil, nil, 500)
	if err != nil {
		return nil, err
	}

	var source *core.Memory
	for i := range mems {
		if mems[i].ID == memoryID {
			source = &mems[i]
			break
		}
	}
	if source == nil {
		return []core.Memory{}, nil
	}

	type scored struct {
		mem       core.Memory
		relevance float64
		reason    string
	}

	candidates := map[string]*scored{}

	// Strategy 1: Shared file references
	fileRefs := extractFileRefs(source.Content)
	if len(fileRefs) > 0 {
		for _, ref := range fileRefs {
			if len(ref) < 3 {
				continue
			}
			hits, _ := c.store.Search(ctx, ref, nil, nil, limit*2)
			for _, hit := range hits {
				if hit.ID != memoryID {
					if existing, ok := candidates[hit.ID]; ok {
						existing.relevance += 0.4
					} else {
						candidates[hit.ID] = &scored{mem: hit, relevance: 0.4, reason: "shared_file"}
					}
				}
			}
		}
	}

	// Strategy 2: Content similarity via FTS search
	queryText := source.Content
	if len(queryText) > 300 {
		queryText = queryText[:300]
	}
	similar, _ := c.store.Search(ctx, queryText, nil, nil, limit*2)
	for _, hit := range similar {
		if hit.ID != memoryID {
			simScore := hit.RelevanceScore
			if simScore == 0 {
				simScore = 0.3
			}
			if existing, ok := candidates[hit.ID]; ok {
				existing.relevance += simScore
			} else {
				candidates[hit.ID] = &scored{mem: hit, relevance: simScore, reason: "content_similar"}
			}
		}
	}

	// Strategy 3: Temporal proximity (same creation window)
	for _, m := range mems {
		if m.ID == memoryID {
			continue
		}
		// Simple temporal proximity: share the same 10-minute window (same timestamp prefix)
		if len(source.CreatedAt) >= 16 && len(m.CreatedAt) >= 16 {
			if source.CreatedAt[:16] == m.CreatedAt[:16] {
				if existing, ok := candidates[m.ID]; ok {
					existing.relevance += 0.2
				} else {
					candidates[m.ID] = &scored{mem: m, relevance: 0.2, reason: "temporal_proximity"}
				}
			}
		}
	}

	// Rank and return
	var ranked []scored
	for _, s := range candidates {
		ranked = append(ranked, *s)
	}
	sort.Slice(ranked, func(i, j int) bool {
		return ranked[i].relevance > ranked[j].relevance
	})

	result := []core.Memory{}
	for i, s := range ranked {
		if i >= limit {
			break
		}
		result = append(result, s.mem)
	}
	return result, nil
}

var fileRefPatterns = []*regexp.Regexp{
	regexp.MustCompile(`[\w/.-]+\.\w{1,5}`),
	regexp.MustCompile(`(?:src|lib|app|tests?)/[\w/.-]+`),
}

func extractFileRefs(text string) []string {
	seen := map[string]bool{}
	var refs []string
	for _, pattern := range fileRefPatterns {
		for _, match := range pattern.FindAllString(text, -1) {
			if len(match) > 5 && !seen[match] && len(match) < 200 {
				seen[match] = true
				refs = append(refs, match)
			}
		}
	}
	if len(refs) > 10 {
		refs = refs[:10]
	}
	return refs
}

// FindSimilarFailures finds failures in other repositories similar to failures in the given repo.
func (c *Correlator) FindSimilarFailures(ctx context.Context, repoID string, limit int) ([]core.Memory, error) {
	// 1. Get recent failures in this repo
	var failures []core.Memory
	mems, err := c.store.ListMemories(ctx, &repoID, nil, 100)
	if err != nil {
		return nil, err
	}
	for _, m := range mems {
		if m.Kind == "failure" {
			failures = append(failures, m)
		}
	}
	if len(failures) == 0 {
		return []core.Memory{}, nil
	}

	// 2. Build a combined query text from recent failures
	var queryText string
	for i, m := range failures {
		if i >= 5 {
			break
		}
		queryText += m.Content + " "
	}
	if len(queryText) > 500 {
		queryText = queryText[:500]
	}

	// 3. Search for failures globally (repoID=nil) using this combined text
	kinds := []string{"failure"}
	similar, err := c.store.Search(ctx, queryText, nil, kinds, limit*2)
	if err != nil {
		return nil, err
	}

	// 4. Filter out failures from the same repo
	var result []core.Memory
	for _, m := range similar {
		if m.RepoID != repoID {
			result = append(result, m)
			if len(result) >= limit {
				break
			}
		}
	}

	return result, nil
}
